src/output-json-modbus.c

```diff
@@ -0,0 +1,355 @@
+{
+    uint16_t val;
+    int i = 0;
+    if (thread->modbuslog_ctx->flags & LOG_MODBUS_USE_DICT) {
+        char key[6];
+        json_t *data_dict = json_object();
+        if (data_dict == NULL)
+            goto end;
+        for (i = 0; i < quantity; i++) {
+            val = ntohs(*(modbustx->data + i));
+            snprintf(key, sizeof(key), "%d", i);
+            json_object_set_new(data_dict, key, json_integer(val));
+        }
+        return data_dict;
+    } else {
+        json_t *data_array = json_array();
+        if (data_array == NULL)
+            goto end;
+        for (i = 0; i < quantity; i++) {
+            val = ntohs(*(modbustx->data + i));
+            json_array_append_new(data_array, json_integer(val));
+        }
+        return data_array;
+    }
+end:
+    return NULL;
+}
+
+static json_t *JsonModbusLogBits(LogModbusLogThread *thread,
+    ModbusTransaction *modbustx, uint8_t quantity)
+{
+    uint16_t val;
+    int i = 0;
+    if (thread->modbuslog_ctx->flags & LOG_MODBUS_USE_DICT) {
+        char key[6];
+        json_t *data_dict = json_object();
+        if (data_dict == NULL)
+            goto end;
+        for (i = 0; i < quantity; i++) {
+            val = *((uint8_t*)modbustx->data + (i >> 3)) & (1 << (i % 8)) ? 1 : 0;
+            snprintf(key, sizeof(key), "%d", i);
+            json_object_set_new(data_dict, key, json_integer(val));
+        }
+        return data_dict;
+    } else {
+        json_t *data_array = json_array();
+        if (data_array == NULL)
+            goto end;
+        for (i = 0; i < quantity; i++) {
+            val = *(modbustx->data + (i >> 3)) & (1 << (i % 8)) ? 1 : 0;
+            json_array_append_new(data_array, json_integer(val));
+        }
+        return data_array;
+    }
+end:
+    return NULL;
+}
+
+
+static int JsonModbusLogger(ThreadVars *tv, void *thread_data,
+    const Packet *p, Flow *f, void *state, void *tx, uint64_t tx_id)
+{
+    ModbusTransaction *modbustx = tx;
+    LogModbusLogThread *thread = thread_data;
+    MemBuffer *buffer = thread->buffer;
+    json_t *js, *modbusjs;
+
+    js = CreateJSONHeader((Packet *)p, 0, "modbus");
+    if (unlikely(js == NULL)) {
+        return TM_ECODE_FAILED;
+    }
+
+    modbusjs = json_object();
+    if (unlikely(modbusjs == NULL)) {
+        goto error;
+    }
+
+    json_object_set_new(modbusjs, "transaction_id", json_integer(modbustx->transactionId));
+    json_object_set_new(modbusjs, "length", json_integer(modbustx->length));
+    json_object_set_new(modbusjs, "function",
+                        json_string(ModbusGetFunctionName(modbustx->function))
+                       );
+    if (modbustx->function == MODBUS_FUNC_DIAGNOSTIC) {
+        json_object_set_new(modbusjs, "subfunction",
+                            json_string(ModbusGetSubFunctionName(modbustx->subFunction))
+                           );
+    }
+    if (modbustx->type & MODBUS_TYP_READ) {
+        json_t *read = json_object();
```

Comment:
read (and write below) are common C function names, please use another name here.